### ISSUE TITLE
How to restrict size of images served based on cookie?

### ISSUE THREAD

**Issue by barryspearce** [cantaloupe-project/cantaloupe#390]
I have a need to restrict images served to a maximum long side of 300 unless a cookie with appropriate content exists.

I can see how to use the context to obtain the cookie, but I do not understand how to restrict the size.
Which part of the delagate script do I need to add the code to?

```
    w = context['full_size']['width']
    h = context['full_size']['height']
    if w > 300 && w > h
        scale = 300.0 / w
        w = 300.0
        h = h * scale
    elsif h > 300
        scale = 300.0 / h
        h = 300.0
        w = w * scale
    end
```

This code does the scaling I am interested in - But I can't work out what to do with the resultant w,h.

Will the system report the new max size in the info.json as well as restricting image requests?

Thanks in advance!

**Comment by adolski** [comment]
Hi @barryspearce,

The IIIF Image API 2.1 doesn't allow a server to return an image other than what the client requests, at the same URI. It must either return what the client requests, or redirect to an alternative, or fail. For example:

```ruby
def authorize(options = {})
  cookie = false # replace with cookie-checking code
  unless cookie
    size = context['resulting_size']
    w    = size['width']
    h    = size['height']
    if w > 300 || h > 300
      uri      = context['request_uri']
      parts    = uri.split('/')
      # this index may differ at your site
      parts[7] = (w > h) ? '300,' : ',300'
      uri      = parts.join('/')
      return {
        'status_code' => 303, # See Other
        'location' => uri
      }
    end
  end
  true
end
```

**Comment by barryspearce** [comment]
Hi @adolski 

ah! Thank you for your help.

I am not sure I am quite there yet as that didn't limit the resolution viewed to the equivalent of a 300px max long side version. I don't think I explained myself properly and I do not understand how OpenSeadragon interacts with the IIIF server to deliver the image wrt zooming etc.

What I am looking to do is Tiered-access. If cookie/bearer header is acceptable full access is allowed, else access is still allowed by the base resolution supplied should be 300 max longside and I don't want this to be circumvented as I am doing this due to copyright restrictions.

Is the correct way to approach this the tiered access example. then working out the scaling constraint? The thing is - I am not worried if the user zooms with upscaling just that the max-base resolution from which the upscaling occurs is based on 300px max longside.

So if we head to a non-zoomed approach traditionally one would supply a 300px long max -side image and a full sized image - which being selected based on the security permission. I am looking to replicate that idea, but obviously using an IIIF server and OSD limiting the base resolution.

I hope that makes sense!
Cheers
Barry

**Comment by adolski** [comment]
> I am not sure I am quite there yet as that didn't limit the resolution viewed to the equivalent of a 300px max long side version. I don't think I explained myself properly and I do not understand how OpenSeadragon interacts with the IIIF server to deliver the image wrt zooming etc.

Roughly the viewer is going to look first at the `info.json` and use the information from that to figure out the size of the initial/smallest image, how many scale factors a.k.a. zoom levels are available, the size of the tiles it should request, etc.

Using scale constraints is the only way with Cantaloupe to affect the `info.json`. So it sounds like your solution would be based on the [tiered access example](https://cantaloupe-project.github.io/manual/4.1/access-control.html#Tiered%20Access) with some extra code to compute a desired scale constraint fraction that gets your "virtual full size" close to 300px.

**Comment by barryspearce** [comment]
Thx for your help - for others who may comes across the same issue and find this, here is what I did using the tiered access example. 


    def authorize(options = {})
        if !context['full_size']
            # Not an image request its a json request.
            return true
        end
        
        if context['cookies'] && context['cookies']['Cookie']
            # Fix cookie handling code until #329 "Cookies header?" is fixed (scheduled milestone 5)
            cookies = Hash[context['cookies']['Cookie'].split(";").map {|str| str.split("=")}]
            cookie = cookies['<YOUR COOOKIE>']
            if cookie == '<YOUR COOKIE VALUE>'  # Replace with appropriate cookie check code.
                return true
            end
        end
        
        # We don't have the required cookie. Turn down to 300 max longest side.
        # First work out the maximum scaling that 300px longest side gves.
        w = context['full_size']['width']
        h = context['full_size']['height']
        
        # If max size is within constraints anyway just OK it...
        if w <= 300 && h <= 300
            return true
        end

        scale = 300.0 / w
        if h > w
            scale = 300.0 / h
        end

        #
        # Get the current scale constraint fraction, which ultimately comes
        # from the identifier URI path component, which is conveniently
        # made available for you in the delegate script context.
        # This is a two-element array of numerator and denominator integers.
        scale_constraint = context['scale_constraint']

        # If it exists, convert it to a Rational (Ruby fraction).
        # Otherwise, use 1/1.
        scale_constraint_r = scale_constraint ? Rational(*scale_constraint) : Rational(1)
        
        # If the requested scale constraint is larger than authorized,
        # redirect to the authorized scale constraint.
        max_auth_scale_r = scale.to_r
        
        if scale_constraint_r > max_auth_scale_r
          return {
              'status_code' => 302,
              'scale_numerator' => max_auth_scale_r.numerator,
              'scale_denominator' => max_auth_scale_r.denominator
          }
        else
          return true
        end
    end

However, Mirador demo is generally happy but sometimes produces some odd visual effects! (Posted below)...some images were ok - others "home" was confused about the full image size etc. Could be a buggy Mirador demo.
![screencapture-projectmirador-org-demo-2020-08-08-23_37_00](https://user-images.githubusercontent.com/18211861/89721056-d0ef7380-d9d0-11ea-87bb-8cfc8fe9a06d.jpg)


But for my use-case where I need the user authenticated to release material > 300px it does the job...and if you request a 300px image and place in an img tag maintaining aspect then everything is perfect.
